File: weyl-core/src/test/java/io/quartic/weyl/core/catalogue/CatalogueWatcherShould.java
package io.quartic.weyl.core.catalogue;

import com.fasterxml.jackson.databind.JavaType;
import io.quartic.catalogue.api.DatasetConfig;
import io.quartic.catalogue.api.DatasetId;
import io.quartic.common.client.WebsocketListener;
import org.junit.Before;
import org.junit.Test;

import java.util.Map;

import static io.quartic.common.test.CollectionUtils.entry;
import static io.quartic.common.test.CollectionUtils.map;
import static io.quartic.common.test.rx.RxUtils.all;
import static io.quartic.weyl.core.catalogue.CatalogueEvent.Type.CREATE;
import static io.quartic.weyl.core.catalogue.CatalogueEvent.Type.DELETE;
import static org.hamcrest.Matchers.contains;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static rx.Observable.just;

public class CatalogueWatcherShould {

    private final WebsocketListener<Map<DatasetId, DatasetConfig>> listener = mock(WebsocketListener.class);
    private final WebsocketListener.Factory listenerFactory = mock(WebsocketListener.Factory.class);

    private final CatalogueWatcher watcher = CatalogueWatcherImpl.of(listenerFactory);

    @Before
    public void before() throws Exception {
        when(listenerFactory.create(any(JavaType.class))).thenReturn((WebsocketListener)listener);
    }

    @Test
    public void emit_creation_event_when_dataset_appears() throws Exception {
        final DatasetId id = mock(DatasetId.class);
        final DatasetConfig config = mock(DatasetConfig.class);

        when(listener.observable()).thenReturn(just(map(entry(id, config))));

        assertThat(all(watcher.events()), contains(CatalogueEventImpl.of(CREATE, id, config)));
    }

    @Test
    public void emit_deletion_event_when_dataset_appears() throws Exception {
        final DatasetId id = mock(DatasetId.class);
        final DatasetConfig config = mock(DatasetConfig.class);

        when(listener.observable()).thenReturn(just(
                map(entry(id, config)),
                map()   // Gone!
        ));

        assertThat(all(watcher.events()), contains(
                CatalogueEventImpl.of(CREATE, id, config),
                CatalogueEventImpl.of(DELETE, id, config)
        ));
    }

    @Test
    public void not_emit_creation_event_when_dataset_persists() throws Exception {
        final DatasetId id = mock(DatasetId.class);
        final DatasetConfig config = mock(DatasetConfig.class);

        when(listener.observable()).thenReturn(just(
                map(entry(id, config)),
                map(entry(id, config))  // Again
        ));

        assertThat(all(watcher.events()), contains(CatalogueEventImpl.of(CREATE, id, config)));
    }

    @Test
    public void emit_events_for_multiple_independent_datasets() throws Exception {
        final DatasetId idA = mock(DatasetId.class);
        final DatasetId idB = mock(DatasetId.class);
        final DatasetConfig configA = mock(DatasetConfig.class);
        final DatasetConfig configB = mock(DatasetConfig.class);

        when(listener.observable()).thenReturn(just(
                map(entry(idA, configA)),
                map(entry(idA, configA), entry(idB, configB)),
                map(entry(idB, configB))
        ));

        assertThat(all(watcher.events()), contains(
                CatalogueEventImpl.of(CREATE, idA, configA),
                CatalogueEventImpl.of(CREATE, idB, configB),
                CatalogueEventImpl.of(DELETE, idA, configA)
        ));
    }
}

File: catalogue-utils/src/test/kotlin/io/quartic/catalogue/CatalogueWatcherShould.kt
package io.quartic.catalogue

import com.fasterxml.jackson.databind.type.MapType
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import io.quartic.catalogue.CatalogueEvent.Type.CREATE
import io.quartic.catalogue.CatalogueEvent.Type.DELETE
import io.quartic.catalogue.api.model.DatasetConfig
import io.quartic.catalogue.api.model.DatasetCoordinates
import io.quartic.catalogue.api.model.DatasetId
import io.quartic.catalogue.api.model.DatasetNamespace
import io.quartic.common.rx.all
import io.quartic.common.websocket.WebsocketListener
import org.hamcrest.Matchers.contains
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import rx.Observable.just

class CatalogueWatcherShould {
    private val listener = mock<WebsocketListener<Map<DatasetNamespace, Map<DatasetId, DatasetConfig>>>>()
    private val listenerFactory = mock<WebsocketListener.Factory>()
    private val ns = mock<DatasetNamespace>()

    private val watcher = CatalogueWatcher(listenerFactory)

    @Before
    fun before() {
        whenever(listenerFactory.create<Map<DatasetNamespace, Map<DatasetId, DatasetConfig>>>(any<MapType>()))
                .thenReturn(listener)
    }

    @Test
    fun emit_creation_event_when_dataset_appears() {
        val id = mock<DatasetId>()
        val config = mock<DatasetConfig>()

        whenever(listener.observable).thenReturn(just(mapOf(ns to mapOf(id to config))))

        assertThat(all(watcher.events), contains(CatalogueEvent(CREATE, coords(ns, id), config)))
    }

    @Test
    fun emit_deletion_event_when_dataset_appears() {
        val id = mock<DatasetId>()
        val config = mock<DatasetConfig>()

        whenever(listener.observable).thenReturn(just(
                mapOf(ns to mapOf(id to config)),
                emptyMap()   // Gone!
        ))

        assertThat(all(watcher.events), contains(
                CatalogueEvent(CREATE, coords(ns, id), config),
                CatalogueEvent(DELETE, coords(ns, id), config)
        ))
    }

    @Test
    fun not_emit_creation_event_when_dataset_persists() {
        val id = mock<DatasetId>()
        val config = mock<DatasetConfig>()

        whenever(listener.observable).thenReturn(just(
                mapOf(ns to mapOf(id to config)),
                mapOf(ns to mapOf(id to config))  // Again
        ))

        assertThat(all(watcher.events), contains(CatalogueEvent(CREATE, coords(ns, id), config)))
    }

    @Test
    fun emit_events_for_multiple_independent_datasets() {
        val idA = mock<DatasetId>()
        val idB = mock<DatasetId>()
        val configA = mock<DatasetConfig>()
        val configB = mock<DatasetConfig>()

        whenever(listener.observable).thenReturn(just(
                mapOf(ns to mapOf(idA to configA)),
                mapOf(ns to mapOf(idA to configA, idB to configB)),
                mapOf(ns to mapOf(idB to configB))
        ))

        assertThat(all(watcher.events), contains(
                CatalogueEvent(CREATE, coords(ns, idA), configA),
                CatalogueEvent(CREATE, coords(ns, idB), configB),
                CatalogueEvent(DELETE, coords(ns, idA), configA)
        ))
    }

    @Test
    fun not_explode_if_specified_namespace_not_present() {
        whenever(listener.observable).thenReturn(just(
                mapOf(mock<DatasetNamespace>() to mapOf(mock<DatasetId>() to mock<DatasetConfig>()))
        ))

        all(watcher.events) // Ensure there's a subscriber
    }

    private fun coords(namespace: DatasetNamespace, id: DatasetId) = DatasetCoordinates(namespace, id)
}
